addons/mail/static/src/model/model_field.js

```diff
@@ -352,103 +352,78 @@ class ModelField {
-                }
-            }
-            return hasChanged;
-        }
-        // not a command
-        return this.set(record, newVal, options);
-    }
-
-    /**
-     * Get the raw value associated to this field. For relations, this means
-     * the local id or list of local ids of records in this relational field.
-     *
-     * @param {mail.model} record
-     * @returns {any}
-     */
-    read(record) {
-        return record.__values[this.fieldName];
-    }
-
-    /**
-     * Set a value on this field. The format of the value comes from business
-     * code.
-     *
-     * @param {mail.model} record
-     * @param {any} newVal
-     * @param {Object} [options]
      * @param {boolean} [options.hasToUpdateInverse] whether updating the
      *  current field should also update its inverse field. Only applies to
      *  relational fields. Typically set to false only during the process of
      *  updating the inverse field itself, to avoid unnecessary recursion.
      * @returns {boolean} whether the value changed for the current field
      */
-    set(record, newVal, options) {
+    parseAndExecuteCommands(record, newVal, options) {
         const currentValue = this.read(record);
-        if (this.fieldType === 'attribute') {
+        if (this.fieldType === 'attribute' && !(newVal instanceof FieldCommand)){
             if (currentValue === newVal) {
                 return false;
             }
             record.__values[this.fieldName] = newVal;
             return true;
         }
-        if (this.fieldType === 'relation') {
-            let hasChanged = false;
-            for (const val of newVal) {
-                switch (val[0]) {
+
+        let hasChanged = false;
+        if (!(newVal instanceof Array)) {
+            newVal = [newVal];
+        }
+        for (const val of newVal) {
+            const name = val instanceof FieldCommand ? val.name : val[0];
+            const data = val instanceof FieldCommand ? val.data: val[1];
+            if (name === 'clear') {
+                hasChanged = this.clear(record, options) || hasChanged;
+            } else if (this.fieldType === 'attribute') {
+                switch (name) {
+                    case 'decrement':
+                        hasChanged = this.parseAndExecuteCommands(record, currentValue - data, options) || hasChanged;
+                        break;
+                    case 'increment':
+                        hasChanged = this.parseAndExecuteCommands(record, currentValue + data, options) || hasChanged;
+                        break;
+                }
+            } else if (this.fieldType === 'relation') {
+                switch (name) {
                     case 'create':
-                        if (this._setRelationCreate(record, val[1], options)) {
-                            hasChanged = true;
-                        }
+                        hasChanged = this._setRelationCreate(record, data, options) || hasChanged;
```

Comment:
Please keep these with the syntax they had before.
The reason to avoid using the `||` notation is because it is very important to realize the method must be called first no matter what, and "hiding" an actual method call (that has side effect!) inside a logical expression is dangerous (in fact, we made the mistake at some point where `hasChanged` was the first operand of the expression and the method was never called).